File: 155-min-stack/min-stack.py
import heapq
class MinStack(object):
    def __init__(self):
        self.minheap=[]
        self.stack=[]
    def push(self, value):
        """
        :type value: int
        :rtype: None
        """
        self.stack.append(value)
        heapq.heappush(self.minheap,value)
    def pop(self):
        """
        :rtype: None
        """
        if not self.stack:
            return
        a=self.stack.pop()
        hp=[]
        while self.minheap and self.minheap[0]!=a:
            b=heapq.heappop(self.minheap)
            hp.append(b)
        if self.minheap:
            heapq.heappop(self.minheap)
        for i in hp:
            heapq.heappush(self.minheap, i)

    def top(self):
        """
        :rtype: int
        """
        return self.stack[-1]

    def getMin(self):
        """
        :rtype: int
        """
        if self.minheap:
            return self.minheap[0]
        return None
        


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(value)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: 155-min-stack/min-stack.cpp
class MinStack {
public:
    stack<int>st;
    priority_queue<int,vector<int>,greater<int>>pq;
    MinStack() {
        
    }
    
    void push(int value) {
        st.push(value);
        pq.push(value);
    }
    
    void pop() {
        int a=st.top();
        st.pop();
        priority_queue<int>buff;
        while(!pq.empty() && (pq.top()!=a)){
            buff.push(pq.top());
            pq.pop();
        }
        pq.pop();
        while(!buff.empty()){
            pq.push(buff.top());
            buff.pop();
        }
    }
    
    int top() {
        if(!st.empty()) return st.top();
        return -1;
    }
    
    int getMin() {
        if(!pq.empty()) return pq.top();
        return -1;
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(value);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */